ligo/types.mligo

```diff
@@ -12,13 +12,65 @@ type balance_int = {x : int ; y : int}
 
 (* Information stored for every initialized tick. *)
 type tick_state = {
+    (* Index of the previous initialized tick.
+        Here we diverge from the article, and effectively store a doubly-linked
+        list of initialized ticks for speed-up
+        (while the article proposes storing a bitmap for this purpose).
+    *)
     prev : tick_index ;
+
+    (* Index of the next initialized tick. *)
     next : tick_index ;
-    delta_liquidity : int ;
+
+    (* Total amount of liquidity change when this tick is crossed bottom to top by i_c.
+
+        TODO: use of this field is opaque. Why do we always subtract it?
+        Where is it initialized to anything except 0? (set_position does not seem to update it)
+        "bottom to top" part is just imagined ATM, it cannot be witnessed from the code yet AFAIU.
+    *)
+    liquidity_delta : int ;
+
+    (* Numbers of positions covering the given tick.
+        Used for garbage collection.
+    *)
     n_positions : nat ;
-    fee_growth_outside : balance_nat ;
+
+    (* Overall number of seconds s_o that the current tick index i_c
+        has spent below (above) this tick by the moment of the last cross with it,
```

Comment:
Regarding this duality, what do you think if we keep a sum type for this field:

```
type seconds_outside
  = Cur_tick_is_below nat  (* ... *)
  | Cur_tick_is_above nat (* ... *)
```

The documentation will lie on this better, and this may save us from some bugs.

We could even do

```
type seconds_outside =
  = Cur_tick_is_below nat  (* ... *)
  | Cur_tick_is_above timestamp (* ... *)
```

which seems more semantically appropriate.

Reply:
> what do you think if we keep a sum type for this field:

Hmm what kind of mistakes would that sum type protect us against?
AFAICT the field will only be used once in `x_to_y_rec`, and in that instance we don't care whether the current tick is above or below the tick being inspected.

> We could even do

Why is one a `nat` and the other a `timestamp`?

Reply:
Also, just noticed `seconds_outside` is updated in `x_to_y_rec` but not in `y_to_x_rec`, which looks like a bug :nabros:

Gonna create a ticket for that as well.

Reply:
> Hmm what kind of mistakes would that sum type protect us against?

We perform "flip" (as it is called in the code comments) of this field every time the current tick `i_c` crosses our tick. And we rely on the fact, that if `i_c` is below our tick, then exactly an even number of flips has occurred. 

But if `i_c` just "touches" our tick when going up and then immediately goes down, we can miss performing the flip a second time, and appear in an inconsistent state. 

However, this does not really matter too much, because as you said, we have only two similar places where crossing must be handled properly: `x_to_y` and `y_to_x`.

> Why is one a nat and the other a timestamp?

Keeping "number of seconds passed since Unix epoch start" in `nat` seems artificial when a dedicated type for this exists. And this way we get slightly complex expressions like `((Tezos.now - epoch_time) - tick.seconds_outside)`.

On the contrary, with the sum type we will get:

```
x_to_y:  // going down
  let seconds_outside = Cur_tick_is_below 
    (match seconds_outside with
      Cur_tick_is_below timestamp -> timestamp
      Cur_tick_is_above secs -> Tezos.now - secs 
    )
```

which probably is not shorter, but at least looks crystal clear and explicit :)

Reply:
> We perform "flip" (as it is called in the code comments) of this field every time the current tick i_c crosses our tick.

Just to double-check we're on the same page, you mean this, right?

```
            (* Flip time growth. *)
            let seconds_outside_new = assert_nat ((Tezos.now - epoch_time) - tick.seconds_outside) in
```

> And we rely on the fact, that if i_c is below our tick, then exactly an even number of flips has occurred.

I confess I'm a bit confused :sweat_smile:  where do we rely on this fact?


> But if i_c just "touches" our tick when going up and then immediately goes down, we can miss performing the flip a second time, and appear in an inconsistent state.

I'm not sure I understand this part either...

In `x_to_y_rec`, we update `seconds_outside` when the new current tick is less than `lo` (the "highest initialized tick lower than or equal to i_c"). In `y_to_x_rec`, we will update `seconds_outside` ([see this issue](https://issues.serokell.io/issue/TCFMM-20)) when the new current tick is greater than or equal to `lo.next`.

In both cases, we want to use the same equation (6.27 from the whitepaper):

```
𝑡𝑜 := 𝑡 − 𝑡𝑜

aka

seconds_outside_new = assert_nat ((Tezos.now - epoch_time) - tick.seconds_outside)
```


How can we miss performing a flip?

> Keeping "number of seconds passed since Unix epoch start" in nat seems artificial when a dedicated type for this exists

But we're not keeping track of that, we're keeping track of the number of seconds that the current index spent on the "outside".

-----

```
x_to_y:  // going down
  let seconds_outside = Cur_tick_is_below 
    (match seconds_outside with
      Cur_tick_is_below timestamp -> timestamp
      Cur_tick_is_above secs -> Tezos.now - secs 
    )
```

I'm not sure I understand this - in `x_to_y`, when we're going down, and we cross some tick `i`, then the previous current tick `ic` will _always_ be above `i`.

So `match seconds_outside with Cur_tick_is_below timestamp` doesn't seem to make sense IMO :thinking:

Reply:
> Taking the discussion above into account, we could update `seconds_outside` as

Hmm I think the only scenario that would cause that check to trip was if we updated `seconds_outside` in `x_to_y` but not in `y_to_x` (or vice-versa). Which is exactly what happened :nabros:

But then this would mean doing the same thing for other fields that have to be updated every time the tick is crossed, like `fee_growth_outside` and `seconds_per_liquidity_outside`.

I'm not opposed to this change, but I am a bit worried that making these types more complex might affect the contract's performance. I remember in skill-game-escrow @pasqu4le was asked to remove a bit of type-safety here and there and apparently that reduced costs (significantly?). I don't have a lot of experience tuning contracts so I'm just throwing it out there :P


> But do you agree that keeping 1628440005 in nat is not very pleasant?

I agree, but I also think that `timestamp` is not suitable either. `timestamp` represents a specific point in time. What we want is a "duration" or "period" type, e.g., something that represents the amount of time passed between two specific points in time (between two `timestamp`s).

It just so happens that `timestamp` is internally implemented as the number of seconds passed since epoch, but that's not relevant. The semantics of the type are "this is a specific point in time".

I'd be ok with something like `type duration = { duration : nat }`.

Reply:
>  I remember in skill-game-escrow @pasqu4le was asked to remove a bit of type-safety here and there and apparently that reduced costs (significantly?)

Oh even so? 

Hm, I see. Probably it's better to create a separate issue and make a quick investigation then.

> It just so happens that timestamp is internally implemented as the number of seconds passed since epoch, but that's not relevant. The semantics of the type are "this is a specific point in time".

Really fair. Despite half of the time `seconds_outside` behaves like a value of `timestamp` time, it's not a point in time, it's a sum of several durations.

And your proposal looks good.

---

Real thanks for this discussion :slightly_smiling_face: 

I created [TCFMM-22](https://github.com/serokell/segmented-cfmm/pull/6) for the further investigation.